js/symbol/collision_tile.js

```diff
@@ -173,56 +175,89 @@ class CollisionTile {
+            return result;
+        }
+
         const collisionBoxArray = this.collisionBoxArray;
         const rotationMatrix = this.rotationMatrix;
-        const anchorPoint = new Point(minX, minY)._matMult(rotationMatrix);
-
-        const queryBox = this.tempCollisionBox;
-        queryBox.anchorX = anchorPoint.x;
-        queryBox.anchorY = anchorPoint.y;
-        queryBox.x1 = 0;
-        queryBox.y1 = 0;
-        queryBox.x2 = maxX - minX;
-        queryBox.y2 = maxY - minY;
-        queryBox.maxScale = scale;
-
-        // maxScale is stored using a Float32. Convert `scale` to the stored Float32 value.
-        scale = queryBox.maxScale;
-
-        const searchBox = [
-            anchorPoint.x + queryBox.x1 / scale,
-            anchorPoint.y + queryBox.y1 / scale * this.yStretch,
-            anchorPoint.x + queryBox.x2 / scale,
-            anchorPoint.y + queryBox.y2 / scale * this.yStretch
-        ];
 
-        const blockingBoxKeys = this.grid.query(searchBox[0], searchBox[1], searchBox[2], searchBox[3]);
-        const blockingBoxKeys2 = this.ignoredGrid.query(searchBox[0], searchBox[1], searchBox[2], searchBox[3]);
-        for (let k = 0; k < blockingBoxKeys2.length; k++) {
-            blockingBoxKeys.push(blockingBoxKeys2[k]);
+        // Generate a rotated geometry out of the original query geometry.
+        // Scale has already been handled by the prior conversions.
+        const rotatedQuery = [];
+        let minX = Infinity;
+        let minY = Infinity;
+        let maxX = -Infinity;
+        let maxY = -Infinity;
+        for (let i = 0; i < queryGeometry.length; i++) {
+            const ring = queryGeometry[i];
+            for (let k = 0; k < ring.length; k++) {
+                const p = ring[k].matMult(rotationMatrix);
+                minX = Math.min(minX, p.x);
+                minY = Math.min(minY, p.y);
+                maxX = Math.max(maxX, p.x);
+                maxY = Math.max(maxY, p.y);
+                rotatedQuery.push(p);
+            }
+        }
+
+        const features = this.grid.query(minX, minY, maxX, maxY);
+        const ignoredFeatures = this.ignoredGrid.query(minX, minY, maxX, maxY);
+        for (let i = 0; i < ignoredFeatures.length; i++) {
+            features.push(ignoredFeatures[i]);
         }
 
-        for (let i = 0; i < blockingBoxKeys.length; i++) {
-            const blocking = collisionBoxArray.get(blockingBoxKeys[i]);
 
+        // Predicate for ruling out already seen features.
+        function seenFeature(sourceLayerFeatures, blocking) {
             const sourceLayer = blocking.sourceLayerIndex;
             const featureIndex = blocking.featureIndex;
             if (sourceLayerFeatures[sourceLayer] === undefined) {
                 sourceLayerFeatures[sourceLayer] = {};
             }
+            return sourceLayerFeatures[sourceLayer][featureIndex] === true;
+        }
```

Comment:
Since this function does not need any variables from the outer scope (it gets everything it needs from arguments), move it out to the top level scope. This way you avoid the cost of setting up a closure on every call of `queryRenderedSymbols`.